wordpress-mobile/WordPress-Android
# Use process lifecycle observer for foreground/background state change detection

Fixes #8890  and introduces LifecycleObserver (see https://developer.android.com/topic/libraries/architecture/lifecycle) as a means to rely on for doing specific tasks when the app comes to the foreground  / goes to the background.

The most relevant part lies in the annotated methods:

``` 
    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    void onAppComesFromBackground() {
        mApplicationLifecycleMonitor.onAppComesFromBackground();
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    void onAppGoesToBackground() {
        mApplicationLifecycleMonitor.onAppGoesToBackground();
    }
``` 

with ` mApplicationLifecycleMonitor`   being nothing else but our own pre-existing private class that keeps handling everything as usual, with minimum changes (split the ApplicationLifecycleMonitor into 2 classes, one helper class that does all the background/foreground processing pretty much as it was, and the memory changes listener has been kept as a new, separate private class that we continue to use normally in the same way (for bitmap cache to get cleared when needed).

These two methods get called when:
- `onAppComesFromBackground`: when _any_ Activity of our app is brought to the foreground
- `onAppGoesToBackground`: a few moments after the _last_ Activity of our app has been sent to the background (read more on [`ProcessLifeCycleOwner`](https://developer.android.com/reference/android/arch/lifecycle/ProcessLifecycleOwner))

#### Considerations:

There were some changes needed as part of removing the previous ActivityLifecycle listener, as now we don't have a reference to the Activity that comes to the foreground / goes to the background. We don't actually need to know which one it was, but some of the logic relied on having a reference to it, its intent and paramaters, for instance ` deferredInit(Activity activity)` needed to receive an Activity context to show a Dialog when checking for Google Services prior to attempting a FCM token registration, or `NotificationsUpdateServiceStarter.startService()` needed a `noteId`  from a specific Activity's Intent to go try fetch the latest body for such a Note.

So even if not strictly related to the background/foreground _detection_  mechanism being changed, the following needed slight modifications as well:

1) Moved the relevant Google Services checking code (and FCM registration if available) into WPMainActivity, as it needs an Activity that has been started, and it is indeed WPMainActivity, given it's the main entry point (we were previously listening to ActivityLifecycle and running deferredInit() for the first activity shown, which in the end is the same thing fwiw). For more info see https://firebase.google.com/docs/cloud-messaging/android/client#sample-play

2) FCM registration was moved from `deferredInit()`. Please note it was _also_ already being called in `updatePushNotificationTokenIfNotLimited()`, so registration of token to backend server is already backed up by that. Important to note, nothing should happen if Google Services are not present - ofc user won't be able to receive push notifications only but code itself won't break. Now moved that specific "first"  call to https://github.com/wordpress-mobile/WordPress-Android/compare/issue/8890-use-process-lifecycle-owner?expand=1#diff-1cf6428dcfe0d429dfdcf7398ef4c056R300.

3) Notifications list updates is now done normally (without looking into the intent for a specific `noteId` extra) in 7f7db1d4e8, as we now don't have an Activity intent to look into with the new scheme (that is, we don't get passed _which_ Activity is coming to the foreground and therefore we don't have the related Intent and parameters making it come on to the foreground).  The Note content update will be done anyway when entering `NotificationsDetailActivity` (see method ` updateUIAndNote`  there). This particular code covered the case of tapping on a push notification whose content in the meantime (between push notification creation and user tapping on it) had already suffered modifications on the server and thus we may be seeing an old copy. This ~will be~ is brought back in ~an upcoming~ PR #8913.


#### To test:
CASE A.1: verify success notifications work again (Media section)
1. start uploading a new media file to the Media section of the app, observe the upload service foreground notification appears, and shows the upload progress
2. send the app to the background
3. when the upload is finished, observe the foreground notification is dismissed and the success notification appears.

CASE A.2: verify success notifications work again (upload within a Post)
1. start a new draft post and insert new media there. Observe the upload sevice foreground notification appears, and shows the upload progress. Exit the Editor.
2. send the app to the background
3. when the upload is finished, observe the foreground notification is dismissed and the success notification appears.

CASE B.1: notifications registration
- verify you can receive push notifications after installing from scratch. You can do this by, after installing and login in to the app, have someone like on a post or comment you wrote, or make a comment on a blog post you wrote. You should receive a push notification related to that event.


Update release notes:
